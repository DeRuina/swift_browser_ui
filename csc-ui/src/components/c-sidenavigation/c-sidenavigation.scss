:host::-webkit-scrollbar {
  width: 12px;
  background: #d8e8ea;
}
:host::-webkit-scrollbar-track {
  border-radius: 10px;
}
:host::-webkit-scrollbar-thumb {
  background: rgba(0, 0, 0, 0.3);
  border-radius: 10px;
  border: solid 3px #d8e8ea;
}

:host(.autoheight) {
  height: calc(100vh - 60px);
  overflow-y: auto;
  overflow-x: hidden;
}

:host(.desktop) {
  flex-grow: 1;
  display: flex;
  flex-shrink: 2;
  flex-basis: 260px;
  max-width: 320px;
  background: var(--csc-primary);
}

.c-sidenavigation {
  background: var(--csc-primary);
  display: flex;
  flex-flow: column nowrap;
  flex: 1;
  gap: 4px;
  min-height: fit-content;
  padding: 24px 0px 24px 24px;
  position: relative;
  transition: transform 0.3s ease;
  width: 100%;
  z-index: 8;

  &__content {
    $content: &;

    display: flex;
    flex-direction: column;
    flex-basis: 260px;
    flex-grow: 1;
    flex-shrink: 2;
    width: 320px;

    &--mobile {
      height: 100vh;
      max-width: 320px;
      overflow-y: scroll;
      position: fixed;
      right: 0;
      top: 0;
      transition: transform 0.2s cubic-bezier(0.25, 0.8, 0.5, 1);
      transform: translateX(0%);
      z-index: 999;

      &#{$content} {
        &--hidden {
          transform: translateX(100%);
        }
      }

      > nav {
        min-height: auto;
        padding-top: 0;
      }
    }
  }

  &__burger {
    background-color: var(--csc-primary);
    display: flex;
    justify-content: flex-end;
    padding: 8px 16px;
  }

  ul {
    display: flex;
    flex-flow: column nowrap;
    gap: 8px;
    list-style: none;
    margin: 0;
    padding: 0;
  }
}

.autoheight > nav {
  height: calc(100vh - 60px);
  overflow-x: hidden;
  overflow-y: auto;
}

.vertical-spacer {
  flex: 1;
  margin-bottom: 8px;
}

.c-overlay {
  background: rgba(0, 0, 0, 0.5);
  bottom: 0;
  left: 0;
  position: fixed;
  right: 0;
  top: 0;
  z-index: 998;
}

@keyframes fadeIn {
  0% {
    background: rgba(0, 0, 0, 0);
  }
  100% {
    background: rgba(0, 0, 0, 0.5);
  }
}

.c-fadeIn {
  animation-duration: 0.5s;
  animation-iteration-count: 1;
  animation-name: fadeIn;
  z-index: 997;
}

c-sidenavigationitem,
c-subnavigationitem {
  display: contents;
}

.visuallyhidden {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}
