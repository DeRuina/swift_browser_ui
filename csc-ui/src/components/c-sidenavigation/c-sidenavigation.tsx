import { mdiArrowRight } from '@mdi/js';
import { Component, Host, h, Prop, Listen, Element } from '@stencil/core';
/**
 * @group Navigation
 */
@Component({
  tag: 'c-sidenavigation',
  styleUrl: 'c-sidenavigation.scss',
  shadow: true,
})
export class CSidenavigation {
  /**
   * Mobile version
   */
  @Prop() mobile: boolean;

  /**
   * Mobile version menu visibility
   */
  @Prop({ mutable: true }) menuVisible: boolean = false; // eslint-disable-line

  @Element() host: HTMLCSidenavigationElement;

  @Listen('itemChange')
  handleChange(event: Event) {
    const slotted = this.host.querySelectorAll('c-sidenavigationitem');
    const target = event.target as HTMLCSidenavigationitemElement;
    const { active } = target;

    slotted.forEach((item) => {
      if (item.querySelector('[slot="subnavitem"]')) {
        item.active = false;
      }
    });

    if (target.querySelector('[slot="subnavitem"]')) {
      target.active = !active;
    } else {
      target.active = true;
    }
  }

  componentDidLoad() {
    const el = document.querySelector('body');

    ['click', 'keyup'].forEach((eventType) => {
      el.addEventListener(eventType, (e: MouseEvent | KeyboardEvent) => {
        if ((e.target as HTMLElement).matches('c-navigationbutton')) {
          if (eventType === 'click') {
            this.menuVisible = !this.menuVisible;
          } else if (e instanceof KeyboardEvent && e.key === 'Enter') {
            this.menuVisible = !this.menuVisible;
          }
        }
      });
    });
  }

  private _closeMenu() {
    this.menuVisible = false;
  }

  render() {
    const classes = {
      'c-sidenavigation': true,
      'hide-menu': !this.menuVisible,
      mobile: !!this.mobile,
      desktop: !this.mobile,
    };

    const containerClasses = {
      'c-sidenavigation__content': true,
      'c-sidenavigation__content--hidden': !this.menuVisible,
      'c-sidenavigation__content--mobile': !!this.mobile,
      'c-sidenavigation__content--desktop': !this.mobile,
    };

    return (
      <Host class={{ desktop: !this.mobile }}>
        <div class={containerClasses}>
          {this.mobile && (
            <div class="c-sidenavigation__burger">
              <c-icon-button inverted text onClick={() => this._closeMenu()}>
                <span class="visuallyhidden">Close sidemenu</span>
                <svg width="24" height="24" viewBox="0 0 24 24">
                  <path d={mdiArrowRight} />
                </svg>
              </c-icon-button>
            </div>
          )}

          <nav class={classes} role="menubar">
            <slot></slot>
            <div class="vertical-spacer"></div>
            <slot name="bottom"></slot>
          </nav>
        </div>

        {this.menuVisible && this.mobile && (
          <div
            class="c-overlay c-fadeIn"
            onClick={() => this._closeMenu()}
          ></div>
        )}
      </Host>
    );
  }
}
