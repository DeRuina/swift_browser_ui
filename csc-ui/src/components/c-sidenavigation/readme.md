# c-sidenavigation



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description                    | Type      | Default     |
| ------------- | -------------- | ------------------------------ | --------- | ----------- |
| `menuVisible` | `menu-visible` | Mobile version menu visibility | `boolean` | `false`     |
| `mobile`      | `mobile`       | Mobile version                 | `boolean` | `undefined` |


## Dependencies

### Depends on

- [c-icon-button](../c-icon-button)

### Graph
```mermaid
graph TD;
  c-sidenavigation --> c-icon-button
  style c-sidenavigation fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
