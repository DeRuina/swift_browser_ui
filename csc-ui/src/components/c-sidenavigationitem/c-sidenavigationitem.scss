:host(.c-sidenavigation-item) {
  align-items: center;
  backface-visibility: hidden;
  background-color: #d8e8ea00;
  border-radius: 4px 0 0 4px;
  color: #fff;
  cursor: pointer;
  display: grid;
  grid-template-columns: 1fr;
  font-weight: 400;
  overflow: hidden;
  position: relative;
  transform: translate3d(0, 0, 0);
  user-select: none;
  min-width: calc(320px - 24px);
}

:host(.c-sidenavigation-item:hover) {
  background-color: #d8e8ea44;
}

:host(.c-sidenavigation-item:focus) {
  outline: none;
}

:host(.c-sidenavigation-item:focus-visible) {
  border-radius: 4px 0 0 4px;
  outline: 2px #fff solid;
  outline-offset: 2px;
}

:host(.c-sidenavigation-item.active) {
  background-color: #d8e8ea;
  color: var(--csc-primary);

  .svg {
    fill: var(--csc-primary);
    transform: rotate(90deg);
  }

  .c-sidenavigation-item__header {
    color: var(--csc-primary);
  }
}

.c-sidenavigation-item {
  &__header {
    align-items: center;
    color: #fff;
    display: grid;
    gap: 8px;
    grid-template-columns: 1fr;
    min-height: 46px;
    padding: 0 12px;

    &--expandable {
      grid-template-columns: auto 1fr;
    }
  }

  &__slot {
    max-width: 100%;
    overflow: hidden;
  }
}

:host(.c-sidenavigation-item--parent.active) {
  padding-bottom: 4px;
}

::slotted(span),
::slotted(c-icon) {
  margin-right: 8px;
  font-size: 20px;
}

::slotted([slot='main']) {
  display: flex;
  align-items: center;
  gap: 6px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  line-height: normal;
}

.styleMain {
  padding-left: 16px;
}

.subnavitem {
  height: 0;
  overflow-y: hidden;
  transition: all 500ms ease;
  width: 100%;
}

.subnavactive {
  height: max-content;
  width: 100%;
}

.svg {
  align-self: center;
  fill: #fff;
  transition: transform 0.3s ease;
}

:host > div.active .svg {
  fill: var(--csc-primary);
  transform: rotate(90deg);
}

.svg.hidden {
  opacity: 0;
}

.middle {
  display: flex;
  height: 100%;
  align-items: center;
}
