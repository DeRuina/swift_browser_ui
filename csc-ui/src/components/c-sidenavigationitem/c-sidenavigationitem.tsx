import {
  Component,
  h,
  Host,
  Prop,
  Element,
  Event,
  EventEmitter,
  Watch,
} from '@stencil/core';
import { mdiChevronRight } from '@mdi/js';

/**
 * @parent c-sidenavigation
 */
@Component({
  tag: 'c-sidenavigationitem',
  styleUrl: 'c-sidenavigationitem.scss',
  shadow: true,
})
export class CSidenavigationitem {
  @Element() hostElement: HTMLCSidenavigationitemElement;

  /**
   * Indicate active state
   */
  @Prop() active: boolean;

  /**
   * Hyperlink url
   */
  @Prop() href: string;

  /**
   * Hyperlink target
   */
  @Prop() target: string = null;

  /**
   * Loading state
   */
  @Prop() loading = false;

  /**
   * Emit changes to the c-accordion
   * @private
   */
  @Event() itemChange: EventEmitter;

  @Watch('active')
  onActiveChange(active: boolean) {
    this._handleChildFocusableChange(active);
  }

  private _ariaLabel: string;

  private _handleChildFocusableChange(focusable: boolean) {
    if (!this._slotHasContent) return;

    const children = Array.from(
      this.hostElement.querySelector('[slot="subnavitem"]').children,
    );

    children.forEach((child: HTMLCSubnavigationitemElement) => {
      child.ariaHidden = (!focusable).toString();
      child.focusable = focusable;
    });
  }

  private _redirect(event: KeyboardEvent | Event) {
    if (
      (event instanceof KeyboardEvent && event?.key === 'Enter') ||
      !(event instanceof KeyboardEvent)
    ) {
      // event.stopPropagation();
      this.itemChange.emit(event);

      if (!this._slotHasContent) {
        const sidenav = document.querySelector('c-sidenavigation');
        sidenav.menuVisible = false;
      }

      if (this.href) {
        if (this.target) {
          window.open(this.href, this.target);
        } else {
          window.location.href = this.href;
        }
      }
    }
  }

  private _slotHasContent = false;

  componentWillLoad() {
    this._slotHasContent = !!this.hostElement.querySelector(
      '[slot="subnavitem"]',
    );

    const children = Array.from(
      this.hostElement.querySelector('[slot="main"]').childNodes,
    );

    this._ariaLabel = children.find((c) => !!c.nodeValue)?.nodeValue?.trim();

    this._handleChildFocusableChange(this.active);
  }

  render() {
    const classes = {
      'c-sidenavigation-item': true,
      'c-sidenavigation-item--parent': this._slotHasContent,
      active: this.active,
    };

    const subNavigationClasses = {
      subnavactive: this.active,
      subnavitem: !this.active,
    };

    const a11y = {
      role: 'menuitem',
      tabindex: '0',
    };

    if (this._slotHasContent) {
      a11y['aria-expanded'] = (!!this.active)?.toString();
    } else if (this.active) {
      a11y['aria-current'] = 'page';
    }

    return (
      <Host
        {...a11y}
        class={classes}
        onClick={(e) => this._redirect(e)}
        onKeyDown={(e) => this._redirect(e)}
      >
        <div
          class={{
            'c-sidenavigation-item__header': true,
            'c-sidenavigation-item__header--expandable': this._slotHasContent,
          }}
        >
          {this._slotHasContent && (
            <svg width="22" height="22" viewBox="0 0 24 24" class="svg">
              <path d={mdiChevronRight} />
            </svg>
          )}
          <div class="c-sidenavigation-item__slot">
            <slot name="main"></slot>
          </div>
        </div>

        {this._slotHasContent && (
          <nav
            role="menubar"
            aria-label={this._ariaLabel}
            aria-expanded={(!!this.active)?.toString()}
            class={subNavigationClasses}
          >
            <slot name="subnavitem"></slot>
          </nav>
        )}

        <c-loader
          size={32}
          hide={!this.loading}
          style={{ pointerEvents: 'none' }}
        ></c-loader>
      </Host>
    );
  }
}
