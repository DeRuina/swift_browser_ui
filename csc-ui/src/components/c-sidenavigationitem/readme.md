# c-sidenavigationitem



<!-- Auto Generated Below -->


## Properties

| Property  | Attribute | Description           | Type      | Default     |
| --------- | --------- | --------------------- | --------- | ----------- |
| `active`  | `active`  | Indicate active state | `boolean` | `undefined` |
| `href`    | `href`    | Hyperlink url         | `string`  | `undefined` |
| `loading` | `loading` | Loading state         | `boolean` | `false`     |
| `target`  | `target`  | Hyperlink target      | `string`  | `null`      |


## Events

| Event        | Description                     | Type               |
| ------------ | ------------------------------- | ------------------ |
| `itemChange` | Emit changes to the c-accordion | `CustomEvent<any>` |


## Dependencies

### Depends on

- [c-loader](../c-loader)

### Graph
```mermaid
graph TD;
  c-sidenavigationitem --> c-loader
  style c-sidenavigationitem fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
