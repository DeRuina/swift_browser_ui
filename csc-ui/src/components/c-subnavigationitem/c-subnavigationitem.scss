.c-subnavigation-item {
  align-items: center;
  background-color: transparent;
  border-radius: 4px;
  color: #595959;
  cursor: pointer;
  display: block;
  display: flex;
  font-weight: 400;
  line-height: 46px;
  margin: 0 8px;
  overflow: hidden;
  padding-left: 34px;
  position: relative;
  transition: background-color 0.2s ease-in;
  user-select: none;
  max-width: calc(320px - 24px - 16px);
}

.c-subnavigation-item__wrapper {
  padding: 2px 0;
}

.c-subnavigation-item__content {
  display: flex;
  align-items: center;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c-subnavigation-item__content::before {
  background-color: var(--csc-primary);
  content: '';
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  transform: translateZ(0) translateX(-8px);
  transition: transform 0.2s ease-in-out;
  width: 8px;
}

.c-subnavigation-item:hover {
  background-color: #f0f6f7;
  color: var(--csc-primary);
}

:host(.active) .c-subnavigation-item {
  background-color: #ffffff;
}

:host(.active) .c-subnavigation-item__content::before {
  transform: translateZ(0) translateX(0);
}

::slotted(span) {
  margin-right: 8px;
  font-size: 20px;
  line-height: 1;
}

.c-subnavigation-item:focus {
  outline: none;
}

.c-subnavigation-item:focus-visible {
  outline: 2px var(--csc-primary) solid;
  outline-offset: 0;
}

.visuallyhidden {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c-subnavigation-item__slot {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;

  slot {
    display: flex;
    align-items: center;
  }
}
