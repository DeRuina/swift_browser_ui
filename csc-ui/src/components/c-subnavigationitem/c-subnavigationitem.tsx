import { Component, h, Host, Element, Prop } from '@stencil/core';
/**
 * @parent c-sidenavigation
 */
@Component({
  tag: 'c-subnavigationitem',
  styleUrl: 'c-subnavigationitem.scss',
  shadow: true,
})
export class CSubnavigationitem {
  @Element() element: HTMLCSubnavigationitemElement;

  /**
   * Active state
   */
  @Prop() active: boolean;

  /**
   * Element is visible and focusable
   */
  @Prop() focusable = false;

  /**
   * Link url
   */
  @Prop() href: string;

  /**
   * Link target
   */
  @Prop() target: string = null;

  /**
   * Loading state
   */
  @Prop() loading = false;

  private _redirect(event: KeyboardEvent | MouseEvent | PointerEvent) {
    if (
      (event instanceof KeyboardEvent && event?.key === 'Enter') ||
      event instanceof MouseEvent ||
      event instanceof PointerEvent
    ) {
      event.stopPropagation();

      const sidenav = document.querySelector('c-sidenavigation');
      sidenav.menuVisible = false;

      if (this.href) {
        if (this.target) {
          window.open(this.href, this.target);
        } else {
          window.location.href = this.href;
        }
      }
    }
  }

  render() {
    const a11y = {
      tabindex: this.focusable ? '0' : '-1',
      role: 'menuitem',
    };

    if (this.active) {
      a11y['aria-current'] = 'page';
    }

    return (
      <Host
        {...a11y}
        class={{ active: this.active }}
        onClick={(e) => this._redirect(e)}
        onKeyDown={(e) => this._redirect(e)}
      >
        <div class="c-subnavigation-item__wrapper">
          <div class="c-subnavigation-item">
            <div class="c-subnavigation-item__content">
              <div class="c-subnavigation-item__slot">
                <slot></slot>
              </div>
              {this.active && (
                <span class="visuallyhidden">, Current page</span>
              )}
            </div>

            <c-loader
              size={32}
              hide={!this.loading}
              style={{ pointerEvents: 'none' }}
            ></c-loader>
          </div>
        </div>
      </Host>
    );
  }
}
