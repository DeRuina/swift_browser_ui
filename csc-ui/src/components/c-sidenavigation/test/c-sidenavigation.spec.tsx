import { newSpecPage } from '@stencil/core/testing';
import { CSidenavigation } from '../c-sidenavigation';
import { CSidenavigationitem } from '../../c-sidenavigationitem/c-sidenavigationitem';
import { CSubnavigationitem } from '../../c-subnavigationitem/c-subnavigationitem';

describe('c-sidenavigation', () => {
  it('renders visible menu', async () => {
    const page = await newSpecPage({
      components: [CSidenavigation, CSidenavigationitem, CSubnavigationitem],
      html: `
      <c-sidenavigation menu-visible=${true}>
        <c-sidenavigationitem active href="google.com" target="_blank">Google</c-sidenavigationitem>
        <c-sidenavigationitem> Something
          <c-subnavigationitem focusable href="somewhere.com" target="_blank"></c-subnavigationitem>
        </c-sidenavigationitem>
        <c-sidenavigationitem href="csc.fi" target="_blank">CSC</c-sidenavigationitem>
      </c-sidenavigation>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
