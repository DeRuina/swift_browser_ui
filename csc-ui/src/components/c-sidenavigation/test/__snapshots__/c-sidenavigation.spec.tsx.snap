// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-sidenavigation renders visible menu 1`] = `
<c-sidenavigation class="desktop" menu-visible="true">
  <mock:shadow-root>
    <div class="c-sidenavigation__content c-sidenavigation__content--desktop">
      <nav class="c-sidenavigation desktop" role="menubar">
        <slot></slot>
        <div class="vertical-spacer"></div>
        <slot name="bottom"></slot>
      </nav>
    </div>
  </mock:shadow-root>
  <c-sidenavigationitem active="" href="google.com" target="_blank">
    Google
  </c-sidenavigationitem>
  <c-sidenavigationitem>
    Something
    <c-subnavigationitem active="" focusable="" href="somewhere.com" target="_blank"></c-subnavigationitem>
  </c-sidenavigationitem>
  <c-sidenavigationitem href="csc.fi" target="_blank">
    CSC
  </c-sidenavigationitem>
</c-sidenavigation>
`;
